### PR TITLE
Add podAffinity and podAntiAffinity to placement spec

### PR DESCRIPTION
**Pull request by galexrt**
This adds `podAffinity` and `podAntiAffinity` to the placement spec.

Example: Setting the `podAntiAffinity` for mons would allow for having one mon per node and not multiple mons on one node.

**NOTE** `podAffinity` and `podAntiAffinity` feature isn't available in Kubernetes 1.5. The compatiblity is guaranteed as long as the specs aren't used/filled out in the rook cluster spec.

### CODE REVIEW

File: pkg/operator/k8sutil/placement.go [hunk 1]
```diff
@@ -36,6 +38,19 @@ func (p Placement) ApplyToPodSpec(t *v1.PodSpec) {
 		}
 		t.Affinity.NodeAffinity = p.NodeAffinity
 	}
+	if p.PodAffinity != nil {
+		if t.Affinity == nil {
```
Comment (galexrt) [inline pkg/operator/k8sutil/placement.go]: If it's wanted we can deduplicate this code by a single if condition before the conditions begin (line 34/5).
Something like this, before line 35:
```
if p.NodeAffinity != nil || p.PodAntiAffinity != nil || p.PodAntiAffinity != nil  {
    t.Affinity = &v1.Affinity{}
}
```

Comment (travisn) [reply pkg/operator/k8sutil/placement.go]: what if we always initialize the Affinity to empty? Is there a side effect when no affinity settings are intended?

File: cluster/examples/kubernetes/rook-cluster.yaml [hunk 2]
```diff
@@ -24,20 +24,30 @@ spec:
...
 #      nodeAffinity:
+#      podAffinity:
+#      podAntiAffinity:
 #      tolerations:
 #    mgr:
 #      nodeAffinity:
+#      podAffinity:
+#      podAntiAffinity:
 #      tolerations:
 #    mon:
```
Comment (travisn) [inline cluster/examples/kubernetes/rook-cluster.yaml]: the mon affinity will need to be handled in another place as well since selection of the mon hosts is explicitly done by the operator. See where the node affinity is implemented: https://github.com/rook/rook/blob/master/pkg/operator/mon/mon.go#L481. The intention of that code is that a mon will be pinned to the node where it lands and won't be rescheduled. I wonder if there is a better way to do that so we don't have to re-implement the affinity handling.

Comment (galexrt) [reply cluster/examples/kubernetes/rook-cluster.yaml]: @travisn The problem I solved for me with adding an anti-affinity was that I had two or more MONs on the same host when one node with a mon restarted and didn't come back up and got failovered.

File: pkg/operator/k8sutil/placement.go (hunk 1 above)
Comment (galexrt) [reply pkg/operator/k8sutil/placement.go]: When updating "old" replicasets they may get updated due to the empty affinity, but I think that Kubernetes is smart enough to see that the change is empty and do nothing.

File: cluster/examples/kubernetes/rook-cluster.yaml (hunk 2 above)
Comment (travisn) [reply cluster/examples/kubernetes/rook-cluster.yaml]: Agreed, the anti-affinity is very valuable to add. Are you saying you tested this change and it worked for mons? I would have expected more changes are necessary for mons because the operator adds a node selector to the mon pods that pins a mon to a node. So if the operator doesn't take the anti-affinity into consideration when it sets the node selector, this could cause issues.

Comment (galexrt) [reply cluster/examples/kubernetes/rook-cluster.yaml]: At least on the minikube one node cluster only one mon was scheduled due to the pod anti-affinity.
A default anti-affinity would add the requirement of having at least three nodes eligible to run a mon on them. But it reduces the single point of failure enormously.

What about modifying the code (in combination to adding a default anti-affinity for mons) to only create a new mon when more eligible nodes are available. I personally would prefer to remove the need of the mon id and only run one mon per eligible node. When one mon fails, the failover should delete the replicaset and reschedule it, as long as the user uses `dataDirHostPath`, a mon always uses the same data path.
If a mon is completely broken and doesn't start anymore, the user should take notice that a (mon) pod isn't running anymore from the metrics that are exposed.

Comment (travisn) [reply cluster/examples/kubernetes/rook-cluster.yaml]: Agreed that it makes sense to run no more than one mon per host in a production environment. For simplicity to test smaller clusters during development we don't want to require the anti-affinity. Good to talk through what design we want for production to make sure it's solid though.

In your proposal are you saying the operator should only help the mons stay healthy on the nodes where they are running and that the admin should be involved in the decision to failover the mon? If you have more than three nodes, why wouldn't you want the operator to do this automatically in case a node dies? 

Today if the `dataDirHostPath` is set and the mon pod dies, the same replicaset will bring up a new pod to replace it, and it will have the same mon id and endpoint. The operator shouldn't need to create a new replicaset. Or did I misunderstand?

Comment (galexrt) [reply cluster/examples/kubernetes/rook-cluster.yaml]: Yes, kind of. When a Mon fails, there is (most of the time) a good reason for it.
I see it like this that the current "fix" for fixing a mon not being in quorum (=~ mon Pod failed/died) is, that it gets "rescheduled" (with a new mon id, after the timeout I added) and we hope that the previous error won't come up again with a "clean" data dir (because of the new mon id).

I would like the operator to do automatic failover, but more in a sense that it doesn't use an already used nodes when going for an approach with one mon per node only (using pod anti affinity or adding a check to the node selection), as written when a mon fails (not just "fails" because I drain'n'reboot the node) but there is a reason behind it and I, as an admin would like to know/investigate why.
That even with three nodes I don't end up with a single point of failure on a node:

node 1 | node 2 | node 3
-------- | ------- | --------
mon 1 | mon 2 + mon 3 | x

The user should be aware of when I only have three nodes for mons, that when one mon is gone it get's dangerous.

***

Let's say I have 4 nodes for mons.

node 1 | node 2 | node 3 | node 4
-------- | ------- | -------- | --------
mon 1 | mon 2 | mon 3 | x

If the mon on node 2 fails (for some other reason than me rebooting the server), I want a new mon to start on node 4 and take over the work of node 2 mon. So it looks like this:

node 1 | node 2 | node 3 | node 4
-------- | ------- | -------- | --------
mon 1 | x | mon 3 | mon 4

Let's say now node 1 fails, then I want node 2 to take over, but not as a mon with a new id, it should look like this:

node 1 | node 2 | node 3 | node 4
-------- | ------- | -------- | --------
x | mon 2 | mon 3 | mon 4

The mon IDs should not increase, they should be a bit like the OSD IDs, only increasing when there is something "new".

Now for the scenario that only a mon (on node 4) fails. I, as an admin, should see as an alert that a pod has been restarted X times in Y time range. Let's say during those pod restarts the mon doesn't connect to the quorum and the "mon out timeout" check failovers the mon (even when the mon is able to join the quorum again in between fails, I as an admin should have alerts as stated to see that a pod that is critical infrastructure restarted for X times).

node 1 | node 2 | node 3 | node 4
-------- | ------- | -------- | --------
mon 1 | mon 2 | mon 3 | x

(The mon IDs stay the same even in case of failover)

***

In the case of only three nodes when one mon fails the `ceph mon_status` should instantly show the issue (when rook is hopefully monitored) and the admin should check the logs and in the "worst" case remove the mon data directory, to bring the mon back up or just remove the pod in K8s and the operator to schedule a new mon on the node with the ID assigned to the node it was on (when there aren't any other nodes eligible to run a mon).

***

To summarize the mon ID part: The operator assigns an ID to each node eligible to run a mon once and keeps the ID assigned as long as the user/admin manually edits a configmap that contains the node->mon ID mapping. With a configmap and the user not having set pod anti affinity (as the operator should already only run one per node if the user wanted it to do so), the user can change the configmap to move mons around the cluster as wanted and even run more than on per node.

This is more a look from my perspective at production setup in general and not dev environment.

Comment (travisn) [reply cluster/examples/kubernetes/rook-cluster.yaml]: The critical design point here is that there is no more than one mon per host. We very much agree that in production you absolutely wouldn't want two mons on the host. 

I don't follow why we would need to assign a mon id permanently to each node. Once a mon is removed from quorum, its id should be considered stale and not be used again. When we start a mon on the node again, a new id can be used.

This PR looks good to merge though. It gives nice flexibility for creating the services.

**Comment by galexrt** [comment]
@travisn Rebased.
